### PR TITLE
New resource `azurerm_iothub_endpoint_cosmosdb_account`

### LINKED ISSUE #20591: CosmosDB endpoint fot Iot Hub
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As you can see **[here](https://techcommunity.microsoft.com/t5/internet-of-things-blog/set-up-an-azure-cosmos-db-custom-endpoint-directly-from-iot-hub/ba-p/3675886)**, Cosmos is supported as a built-in endpoint for Iot Hub, but:

- There is no "azurerm_iothub_endpoint_cosmosdb_account" resource

- There is no Cosmos string on the "type" key of the "endpoint" block, for the "azurerm_iothub" resource:
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/iothub#endpoint

 

### New or Affected Resource(s)/Data Source(s)

azurerm_iothub_endpoint_cosmosdb_account

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by myc2h6o**
Close #20591

### CODE REVIEW

**Comment by myc2h6o** [comment]
Test result
image

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for this @myc2h6o - this is looking pretty good, I've left a couple of minor comments but can have another look once those get fixed up. Thanks!

File: internal/services/iothub/iothub_endpoint_cosmosdb_account_resource.go [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+				if strings.EqualFold(pointer.From(endpoint.Name), id.EndpointName) {
+					authenticationType := devices.AuthenticationType(state.AuthenticationType)
+
+					if authenticationType == devices.AuthenticationTypeKeyBased {
+						if state.PrimaryKey == "" || state.SecondaryKey == "" {
+							return fmt.Errorf("`primary_key` and `secondary_key` must be specified when `authentication_type` is `keyBased`")
+						}
+
+						endpoint.PrimaryKey = pointer.To(state.PrimaryKey)
+						endpoint.SecondaryKey = pointer.To(state.PrimaryKey)
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply internal/services/iothub/iothub_endpoint_cosmosdb_account_resource.go]: thanks for pointing this out! I've fixed it

File: internal/services/iothub/iothub_endpoint_cosmosdb_account_resource.go [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+				return err
+			}
+
+			locks.ByName(id.IotHubName, IothubResourceName)
+			defer locks.UnlockByName(id.IotHubName, IothubResourceName)
+
+			iothub, err := client.Get(ctx, id.ResourceGroup, id.IotHubName)
+			if err != nil {
+				if utils.ResponseWasNotFound(iothub.Response) {
+					return nil
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply internal/services/iothub/iothub_endpoint_cosmosdb_account_resource.go]: have added the error return

**Review by catriona-m — APPROVED** [review]
Thanks for updating @myc2h6o - LGTM 🎷

**Pull request merged**